apis/projectcontour/v1/helpers_test.go

```diff
@@ -0,0 +1,433 @@
+						Message: "We had a straightforward error",
+						Status:  ConditionTrue,
+					},
+					{
+						Type:    "SecondTest",
+						Reason:  "TestReason2",
+						Message: "We had an extra straightforward error",
+						Status:  ConditionTrue,
+					},
+				},
+			},
+		},
+		"same reason, multiple type": {
+			dc: &DetailedCondition{
+				Condition: Condition{
+					Type: "Valid",
+				},
+			},
+			subconditions: []subConditionDetails{
+				{
+					condType: "SimpleTest",
+					reason:   "TestReason",
+					message:  "We had a straightforward error",
+				},
+				{
+					condType: "SecondTest",
+					reason:   "TestReason",
+					message:  "We had an extra straightforward error",
+				},
+			},
+			want: &DetailedCondition{
+				Condition: Condition{
+					Type:    "Valid",
+					Status:  ConditionFalse,
+					Reason:  "MultipleProblems",
+					Message: "Multiple problems were found, see errors for details",
+				},
+				Errors: []SubCondition{
+					{
+						Type:    "SimpleTest",
+						Reason:  "TestReason",
+						Message: "We had a straightforward error",
+						Status:  ConditionTrue,
+					},
+					{
+						Type:    "SecondTest",
+						Reason:  "TestReason",
+						Message: "We had an extra straightforward error",
+						Status:  ConditionTrue,
+					},
+				},
+			},
+		},
+		"same reason, same type": {
+			dc: &DetailedCondition{
+				Condition: Condition{
+					Type: "Valid",
+				},
+			},
+			subconditions: []subConditionDetails{
+				{
+					condType: "SimpleTest",
+					reason:   "TestReason",
+					message:  "We had a straightforward error",
+				},
+				{
+					condType: "SimpleTest",
+					reason:   "TestReason",
+					message:  "We had an extra straightforward error",
+				},
+			},
+			want: &DetailedCondition{
+				Condition: Condition{
+					Type:    "Valid",
+					Status:  ConditionFalse,
+					Reason:  "SimpleTestTestReason",
+					Message: "We had a straightforward error, We had an extra straightforward error",
+				},
+				Errors: []SubCondition{
+					{
+						Type:    "SimpleTest",
+						Reason:  "TestReason",
+						Message: "We had a straightforward error, We had an extra straightforward error",
+						Status:  ConditionTrue,
+					},
+				},
+			},
+		},
+		"multiple different reason, same type": {
+			dc: &DetailedCondition{
+				Condition: Condition{
+					Type: "Valid",
+				},
+			},
+			subconditions: []subConditionDetails{
+				{
+					condType: "SimpleTest",
+					reason:   "TestReason",
+					message:  "We had a straightforward error",
+				},
+				{
+					condType: "SimpleTest",
+					reason:   "TestReason2",
+					message:  "We had an extra straightforward error",
+				},
+			},
+			want: &DetailedCondition{
+				Condition: Condition{
+					Type:    "Valid",
+					Status:  ConditionFalse,
+					Reason:  "MultipleProblems",
+					Message: "Multiple problems were found, see errors for details",
+				},
+				Errors: []SubCondition{
+					{
+						Type:    "SimpleTest",
+						Reason:  "MultipleReasons",
```

Comment:
so in this case we lose `TestReason` and `TestReason2` - is it possible to retain them, either in the Reason or Message field?

Reply:
That's a good point, yes, we shouldn't lose them altogether. I'll have a think about where to put them. Possibly putting them into the message concatenation might work.

Reply:
Actually, I think this is okay, because the `Reason` and `Message` fields are supposed to hold the same information, the message is just in a more readable format. I think that `MultipleReasons` is probably more understandable than `TestReasonTestReason2` in this case.

In practice, I don't think this should happen all that often.